db/changes.go

```diff
@@ -171,6 +172,160 @@ func (db *Database) AddDocInstanceToChangeEntry(entry *ChangeEntry, doc *Documen
+		}
+	}
+
+	go func() {
+		defer base.FatalPanicHandler()
+		defer close(feed)
+		var itemsSent int
+		var lastSeq uint64
+
+		for {
+			if requestLimit == 0 {
+				paginationOptions.Limit = queryLimit
+			} else {
+				remainingLimit := requestLimit - itemsSent
+				paginationOptions.Limit = base.MinInt(remainingLimit, queryLimit)
+			}
+
+			// Get changes from 0 to latest seq
+			base.TracefCtx(db.Ctx, base.KeyChanges, "Querying channel %q for revocations with options: %+v", base.UD(singleChannelCache.ChannelName()), paginationOptions)
+			changes, err := singleChannelCache.GetChanges(paginationOptions)
+			if err != nil {
+				base.WarnfCtx(db.Ctx, "Error retrieving changes for channel %q: %v", base.UD(singleChannelCache.ChannelName()), err)
+				change := ChangeEntry{
+					Err: base.ErrChannelFeed,
+				}
+				feed <- &change
+				return
+			}
+			base.DebugfCtx(db.Ctx, base.KeyChanges, "[changesFeed] Found %d changes for channel %q", len(changes), base.UD(singleChannelCache.ChannelName()))
+
+			sentChanges := 0
+			for _, logEntry := range changes {
+				seqID := SequenceID{
+					Seq:         logEntry.Sequence,
+					TriggeredBy: triggeredBy,
+				}
+
+				idOnlyRevocation := false
+
+				// We need to check whether a change / document sequence is greater than since.
+				// If its less than we can continue. Otherwise:
+				// Because we only see the active revision of the document through the GetChanges query, we need to check
+				// whether the user had access to this document at the last since to know whether we need to  end a
+				// revocation message. We don't want to send a revocation message if the document has not seen before.
+				if logEntry.Sequence > sinceVal {
+					requiresRevocation, err := db.docRequiresRevocation(logEntry.DocID, singleChannelCache.ChannelName(), sinceVal)
+					if err != nil {
+						change := ChangeEntry{
+							Err: base.ErrChannelFeed,
+						}
+						feed <- &change
+						return
+					}
+
+					if !requiresRevocation {
+						return
+					}
+
+					idOnlyRevocation = true
+				}
+
+				change := makeRevocationChangeEntry(logEntry, seqID, singleChannelCache.ChannelName(), idOnlyRevocation)
+				lastSeq = logEntry.Sequence
+
+				base.DebugfCtx(db.Ctx, base.KeyChanges, "Channel feed processing seq:%v in channel %s", seqID, base.UD(singleChannelCache.ChannelName()))
+
+				select {
+				case <-options.Terminator:
+					base.DebugfCtx(db.Ctx, base.KeyChanges, "Revoked thing stopped")
+					return
+
+				case feed <- &change:
+					sentChanges++
+				}
+			}
+
+			if len(changes) < paginationOptions.Limit {
+				return
+			}
+
+			itemsSent += sentChanges
+			if requestLimit > 0 && itemsSent >= requestLimit {
+				return
+			}
+
+			paginationOptions.Since.Seq = lastSeq
+		}
+	}()
+
+	return feed
+}
+
+// Checks if a document needs to be revoked. This is used in the case where the since < doc sequence
+func (db *Database) docRequiresRevocation(docID, chanName string, since uint64) (bool, error) {
```

Comment:
I feel like this might be better named for what it's doing (something like wasDocInChannelAtSeq), instead of how we're using the results (revocation check).

Reply:
Yeah that makes sense